### PR TITLE
Tests template

### PR DESCRIPTION
**Pull request by hussainmansour**

### CODE REVIEW

**Review by EhabYasser25 — APPROVED** [review]

File: commands_function.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+class Icommand(metaclass=ABCMeta):
+    @abstractstaticmethod
+    def excute():
+        """the interface Icommand"""
+
+class commandfactory:
+    @staticmethod
+    def build_command(args):
+        command_type=args.command
```
Comment (bangkook) [inline commands_function.py]: add space before and after the '=' operator

File: commands_function.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        schema=self.schema
+        f=open(schema,"r")
+        schima=json.load(f)
+        parent_dir=os.getcwd() #get the current dir of the project
+        data_base_name=schima[keys.DB_name]
+        path=os.path.join(parent_dir,data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path)
+        for table in schima[keys.tables]:
+            table_path=os.path.join(path,table[keys.name])
```
Comment (bangkook) [inline commands_function.py]: same previous comment for all assignments

File: commands_function.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            return get()
+        elif command_type == "set":
+            return set()
+        
+class create(Icommand):
+    def __init__(self,schema):
+        self.schema=schema
+    def excute(self):
+        schema=self.schema
+        f=open(schema,"r")
```
Comment (bangkook) [inline commands_function.py]: don't name a variable 'f' only

File: main.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import Parser
+import commands_function
+
+
+if __name__=='__main__':
+    args=Parser.parseInput()
+    command =commands_function.commandfactory.build_command(args)
```
Comment (bangkook) [inline main.py]: you can use
"from commands_function import commandfactory"
to call commandfactory directly without referring to the file name

Comment (bangkook) [reply main.py]: same with Parser

Comment (josephShokry) [reply main.py]: this is a good one, I didn't know about it before.
I will.

File: commands_function.py (hunk 2 above)
Comment (josephShokry) [reply commands_function.py]: ok!

File: commands_function.py (hunk 3 above)
Comment (josephShokry) [reply commands_function.py]: thanks for the mention

File: commands_function.py (hunk 1 above)
Comment (josephShokry) [reply commands_function.py]: ok we will

File: Parser.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import argparse
+import os
+
+parser = argparse.ArgumentParser()
+parent_path=os.getcwd()
```
Comment (mohamedabdelsalam) [inline Parser.py]: Why do we need this ?

File: Parser.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+def parseInput():
+    parser=argparse.ArgumentParser()
+    parser.add_argument("-cmd","--command",type=str.lower,help="the quiry command",required=True,choices=["get","set","create","delete"])
```
Comment (mohamedabdelsalam) [inline Parser.py]: follow convension, space after `,`

File: Parser.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    parser.add_argument("-sch","--schema",type=str,required=(parser.parse_known_args()[0].command=="create"),help="path to your schema file")
```
Comment (mohamedabdelsalam) [inline Parser.py]: space after `=` and before in all lines

File: commands_function.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+from abc import ABCMeta, abstractstaticmethod
+import json
+from keys import keys
+
+class Icommand(metaclass=ABCMeta):
+    @abstractstaticmethod
+    def excute():
+        """the interface Icommand"""
+
+class commandfactory:
```
Comment (mohamedabdelsalam) [inline commands_function.py]: file name should be `command_factory` not `commands_factory`

File: commands_function.py (hunk 1 above)
Comment (mohamedabdelsalam) [reply commands_function.py]: You are create command not build it. Because there is two design battern regarding to object creation :- 
1- Factory design pattern 
2- Builder design pattern. 

Here, you are creating the object, not building it

Comment (mohamedabdelsalam) [inline commands_function.py]: lower case to the command_type

File: commands_function.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        if command_type == "create":
+            return create(args.schema)
+        elif command_type == "delete":
+            return delete()
+        elif command_type == "get":
+            return get()
+        elif command_type == "set":
+            return set()
+        
+class create(Icommand):
```
Comment (mohamedabdelsalam) [inline commands_function.py]: name it as `CreateCommand` not a Create. + class name should start with Capital

File: commands_function.py (hunk 3 above)
Comment (mohamedabdelsalam) [reply commands_function.py]: +1

File: commands_function.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self,schema):
```
Comment (mohamedabdelsalam) [inline commands_function.py]: space after `,`

File: commands_function.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        self.schema=schema
```
Comment (mohamedabdelsalam) [inline commands_function.py]: space after `=`

Comment (mohamedabdelsalam) [inline commands_function.py]: Is this schema or schema_file ? it is confusing name

File: commands_function.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    def excute(self):
+        schema=self.schema
+        f=open(schema,"r")
+        schima=json.load(f)
```
Comment (mohamedabdelsalam) [inline commands_function.py]: schema not schima 

What happend if the file doesn't exists ? 

I would recommend create a private method here, this method takes a schema_file and return schema object

File: commands_function.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        parent_dir=os.getcwd() #get the current dir of the project
+        data_base_name=schima[keys.DB_name]
+        path=os.path.join(parent_dir,data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path)
```
Comment (mohamedabdelsalam) [inline commands_function.py]: again, why don't u use mkdir with boolean flag, you don't need to use isDir

File: keys.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
+class keys:
```
Comment (mohamedabdelsalam) [inline keys.py]: class name is confusing. 
fllow the convension as well 

- constants in the class should be capital, e.g SCHEMA, TABLES, etc 

- Also search how to make sure in python they are constant

File: schema.txt [hunk 15]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (mohamedabdelsalam) [inline schema.txt:1]: Why you didn't create a folder call tests and move it there ?

File: commands_function.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
+import os
```
Comment (mohamedabdelsalam) [inline commands_function.py]: Create file called ICommand 
file called command_factory
and file for each command

File: test_cases.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import unittest
+import os
+from keys import keys
+import json
+
+parent_path=os.getcwd()
+
+class test_functions(unittest.TestCase):
```
Comment (mohamedabdelsalam) [inline test_cases.py]: You are not handling al test cases, please cover more cases and more edge cases

File: Parser.py (hunk 5 above)
Comment (josephShokry) [reply Parser.py]: sorry I forgot to delete them

File: Parser.py (hunk 6 above)
Comment (josephShokry) [reply Parser.py]: done!

File: commands_function.py (hunk 8 above)
Comment (josephShokry) [reply commands_function.py]: ok done

File: commands_function.py (hunk 1 above)
Comment (josephShokry) [reply commands_function.py]: @mohamedabdelsalam sorry can you explain more.

File: commands_function.py (hunk 13 above)
Comment (josephShokry) [reply commands_function.py]: ok, but I don't know what is wrong with the isDir function.

File: keys.py (hunk 14 above)
Comment (josephShokry) [reply keys.py]: ok did it

File: schema.txt (hunk 15 above)
Comment (josephShokry) [reply schema.txt:1]: I did but did not push it to GitHub
and did not put this file in it because it is an essential file 
but after mentioning it and thinking again you are right and I will.

File: Parser.py (hunk 7 above)
Comment (josephShokry) [reply Parser.py]: did it.

File: Parser.py [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    parser.add_argument("-sch", "--schema", type = str, required = (parser.parse_known_args()[0].command == "create"), help = "path to your schema file")
+    parser.add_argument("-pk", "--primaryKey", type = str, help = "the primary key of the DB row you want to access")
+    parser.add_argument("-val", "--value", type = str, help = "the value that you want to change or add to the DB")  
+    parser.add_argument("-db", "--data_base", type = str, help = "the database name")
+    parser.add_argument("-tb", "--table", type = str, help = "the table name that you want to access")  
+    
+
+    return parser.parse_args()  
+    
+
```
Comment (mohamedabdelsalam) [inline Parser.py]: extra vertical spaces

File: Parser.py [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+def parseInput():
+    parser=argparse.ArgumentParser()
+    parser.add_argument("-cmd", "--command", type = str.lower, help = "the quiry command", required = True, choices = ["get", "set", "create", "delete"])
+    parser.add_argument("-sch", "--schema", type = str, required = (parser.parse_known_args()[0].command == "create"), help = "path to your schema file")
+    parser.add_argument("-pk", "--primaryKey", type = str, help = "the primary key of the DB row you want to access")
+    parser.add_argument("-val", "--value", type = str, help = "the value that you want to change or add to the DB")  
+    parser.add_argument("-db", "--data_base", type = str, help = "the database name")
+    parser.add_argument("-tb", "--table", type = str, help = "the table name that you want to access")  
+    
+
```
Comment (mohamedabdelsalam) [inline Parser.py]: extra vertical spaces

File: command_function.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    @abstractstaticmethod
+    def excute():
+        """the interface Icommand"""
+
+class commandfactory:
+    @staticmethod
+    def build_command(args):
+        command_type=args.command
+        if command_type == "create":
+            return create(args.schema)
```
Comment (mohamedabdelsalam) [inline command_function.py]: Is this schema or schema_file ?

File: command_function.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        elif command_type == "delete":
+            return delete()
+        elif command_type == "get":
+            return get()
+        elif command_type == "set":
+            return set()
+        
+class create(Icommand):
```
Comment (mohamedabdelsalam) [inline command_function.py]: - class name should start with Capital letter 
- make every class in a file
- put al commands classes in one folder

File: command_function.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self, schema):
+        self.schema=schema
+    def excute(self):
+        schema = self.schema
+        schema_file = open(schema, "r")
+        schema = json.load(schema_file)
```
Comment (mohamedabdelsalam) [inline command_function.py]: move this in constucrtor

File: command_function.py [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema[keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
```
Comment (mohamedabdelsalam) [inline command_function.py]: use mkdir(path, true)

File: command_function.py [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+            os.mkdir(path)
+        for table in schema[keys.TABLES]:
+            table_path = os.path.join(path, table[keys.NAME])
+            if not os.path.isdir(table_path):
+                os.mkdir(table_path)
+                # file_name=os.path.join(table_path, "info.txt")
```
Comment (mohamedabdelsalam) [inline command_function.py]: why do we have this comments ?

File: command_function.py [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+        schema_file = open(schema, "r")
+        schema = json.load(schema_file)
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema[keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path)
+        for table in schema[keys.TABLES]:
+            table_path = os.path.join(path, table[keys.NAME])
+            if not os.path.isdir(table_path):
```
Comment (mohamedabdelsalam) [inline command_function.py]: use mkdir(path, true)

File: schema_keys.py [hunk 26]
```diff
@@ -0,0 +1,24 @@
+from imp import PKG_DIRECTORY
+from msilib import schema
+from tkinter.tix import COLUMN
+
+
+class keys:
```
Comment (mohamedabdelsalam) [inline schema_keys.py]: class name capital

File: schema_keys.py [hunk 27]
```diff
@@ -0,0 +1,24 @@
+from imp import PKG_DIRECTORY
+from msilib import schema
+from tkinter.tix import COLUMN
+
```
Comment (mohamedabdelsalam) [inline schema_keys.py]: Why do we need those imports ?

File: Parser.py [hunk 28]
```diff
@@ -0,0 +1,17 @@
+import argparse
+import os
+
+
```
Comment (mohamedabdelsalam) [inline Parser.py]: extra vertical spaces

File: command_function.py (hunk 24 above)
Comment (josephShokry) [reply command_function.py]: this is the code used to make table schema details (the names of columns, the name of the primary key, . . )
I am not sure if we really need it in the future and in what format I will need this file so I made this comment in case I need it in the future I can remove the comment on these lines and edit it.

File: command_function.py (hunk 23 above)
Comment (EhabYasser25) [reply command_function.py]: I dont understand exactly, do I need to use mkdir(path, true) instead of os.mkdir(path)?

File: command_function.py (hunk 21 above)
Comment (EhabYasser25) [reply command_function.py]: Ok, I will handle this.

File: CommandsFunctions/CommandFactory.py [hunk 29]
```diff
@@ -0,0 +1,17 @@
+from Create import Create
+from Delete import Delete
+from Get import Get
+from Set import Set
+
+class CommandFactory:
+    @staticmethod
+    def build_command(args):
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/CommandFactory.py:8]: it is not build, it create

File: CommandsFunctions/CommandFactory.py [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+        command_type=args.command
+        if command_type == "create":
+            return Create(args.schema_file)
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/CommandFactory.py:11]: `CreateCommand` not `Create`. class name can't be a verb

File: CommandsFunctions/Create.py [hunk 31]
```diff
@@ -0,0 +1,36 @@
+import os
+import json
+from SchemaKeys import Keys
+from Icommand import Icommand
+
+class Create(Icommand):
+    def __init__(self, schema_file):
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Create.py:7]: too much try and catch isn't good. 
it is better to have validate function. It throws exception in case anything error, then execute your code

Comment (mohamedabdelsalam) [reply CommandsFunctions/Create.py:7]: look at @bangkook code as an example. I did a comment in this part there

File: CommandsFunctions/Create.py [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+            print('schema is written in a wrong json format')
+            exit()
+        schema.close()
+    def execute(self):
+        schema_data = self.schema_data
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema_data[Keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path, True)
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Create.py:26]: you use True flag, why do you need to check ?

File: CommandsFunctions/Create.py [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+        try:
+            for table in schema_data[Keys.TABLES]:
+                table_path = os.path.join(path, table[Keys.NAME])
+                if not os.path.isdir(table_path):
+                    os.mkdir(table_path, True)
+                    # file_name = os.path.join(table_path, "info.txt")
+                    # f = open(file_name, 'w')
+                    # f.write(json.dumps(i,  indent = 4))
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Create.py:34]: remove commnets

File: CommandsFunctions/Create.py [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+        schema_data = self.schema_data
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema_data[Keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path, True)
+        try:
+            for table in schema_data[Keys.TABLES]:
+                table_path = os.path.join(path, table[Keys.NAME])
+                if not os.path.isdir(table_path):
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Create.py:30]: no need for if condition here

File: CommandsFunctions/Create.py [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+                    os.mkdir(table_path, True)
+                    # file_name = os.path.join(table_path, "info.txt")
+                    # f = open(file_name, 'w')
+                    # f.write(json.dumps(i,  indent = 4))
+        except:
+            print('No Tables Found')
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Create.py:36]: Don't print inside the class

File: CommandsFunctions/Get.py [hunk 36]
```diff
@@ -0,0 +1,8 @@
+import os
+import json
+from SchemaKeys import Keys
+from Icommand import Icommand
+
+class Get(Icommand):
+    def execute(self):
+        print("dummy get will be implemented later")
```
Comment (mohamedabdelsalam) [inline CommandsFunctions/Get.py:8]: raise error

File: SchemaKeys.py [hunk 37]
```diff
@@ -0,0 +1,13 @@
+from imp import PKG_DIRECTORY
+from msilib import schema
+from tkinter.tix import COLUMN
+
+class Keys:
```
Comment (mohamedabdelsalam) [inline SchemaKeys.py]: This class should be abstract

File: TestCases.py [hunk 38]
```diff
@@ -0,0 +1,22 @@
+import unittest
```
Comment (mohamedabdelsalam) [inline TestCases.py:15]: write more unit test to test failure senatios
